src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.cs

```diff
@@ -0,0 +1,170 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json.Serialization
+{
+    public static partial class JsonConverter
+    {
+        private static readonly JsonConverterSettings s_default_options = new JsonConverterSettings();
+
+        private static bool ReadData(
+            ref Utf8JsonReader reader,
+            JsonConverterSettings options,
+            Type returnType,
+            ref JsonObjectState current,
+            ref List<JsonObjectState> previous,
+            ref int arrayIndex)
+        {
+            while (reader.Read())
+            {
+                JsonTokenType tokenType = reader.TokenType;
+                if (tokenType == JsonTokenType.String)
+                {
+                    CheckForNonObjectReturn(ref current, options, returnType);
+                    if (HandleString(ref reader, options, returnType, ref current))
+                    {
+                        return true;
+                    }
+                }
+                else if (tokenType == JsonTokenType.Number)
+                {
+                    CheckForNonObjectReturn(ref current, options, returnType);
+                    if (HandleNumber(ref reader, options, returnType, ref current))
+                    {
+                        return true;
+                    }
+                }
+                else if (tokenType == JsonTokenType.PropertyName)
+                {
+                    Debug.Assert(current.ReturnValue != default);
+                    Debug.Assert(current.ClassInfo != default);
+
+                    ReadOnlySpan<byte> propertyName = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan;
+                    current.PropertyInfo = current.ClassInfo.GetProperty(current.ReturnValue.GetType(), propertyName, current.PropertyIndex);
+                    current.PropertyIndex++;
+                }
+                else if (tokenType == JsonTokenType.StartObject)
+                {
+                    HandleStartObject(options, returnType, ref current, ref previous, ref arrayIndex);
```

Comment:
Why don't we need to call `CheckForNonObjectReturn` for this case? We do it for all the others.

Reply:
Because its an Object :) ... The check is to initialize the "class" that has the metadata for the deseserialize and when returning primitives and arrays that isn't set up yet.